### ISSUE TITLE
Conflict with jax weight initialization and klax.nn.Linear

### ISSUE THREAD

**Issue by Drenderer** [Drenderer/klax#25]
The weight initializers in `jax.nn.initializers` (e.g., he_uniform, he_normal, glorot_uniform, ...), per default, use the second last axis of the weight matrix as the input axis and the last axis as the output axis. However, in the klax.nn.Linear implementation the `@` operator is used for matrix-vector multiplication, which uses the input-output axes definitions from normal mathematics. I think jax expects the weight multiplication like this: `jnp.matmul(x, weight)` which is the transpose of what klax is currently doing. This means that the klax implementation is currently inconsistent with the jax initializers. This should either be fixed by using the transposed matrix multiplications in `__call__` (I find this confusing), by calling the initializers transposed, i.e., `weight = initializer(key, shape=(in_features_, out_features_)).T` or by at least warning the user explicitly about this behavior, so they can change the default in/out axis of their initializers.

**Comment by jaosch** [comment]
In line 70 klax.nn._linear.py you see that the weights are actually initialized with `whape = (out_feature, in_features)`. Does this change anything about your observation?

**Comment by Drenderer** [comment]
It is in a certain sense the origin of the issue. The default initializers from jax expect `whape = (in_features, out_feature)`.

**Comment by Drenderer** [comment]
This issue is addressed in the current version of the feature/nn-linear branch. In the linear layers, the weight is now multiplied $W$ with the input $x$ *from the left* $x^\intercal W$. As a result, the layers are now compatible with the default values of the initializer `in_axes` and `out_axes`. One slight disadvantage remains: When manually inspecting the weights, they will now appear transposed, i.e., for `in_features=3` and `out_features=2`, the weight shape is now `(3, 2)` instead of `(2,3)`.
However, I think this is the lesser evil, since our other options are:
- Transpose the weight `shape` before passing it to the initializer and then transpose the initializer output back. This creates the issue that the arguments `in_axes` and `out_axes` of the initializer do no longer correspond to `in_features` and `out_features`.
- Wrap the JAX initializers in a custom module to manipulate their default values.

**Comment by Drenderer** [comment]
With the current changes to the linear layers in feature/nn-linear implementing the left multiplication, this issue will be closed as soon as the feature is merged into main.
